server/src/main/java/org/elasticsearch/action/bulk/SimulateBulkRequest.java

```diff
@@ -84,4 +128,41 @@ public Map<String, Map<String, Object>> getPipelineSubstitutions() {
     public boolean isSimulated() {
         return true;
     }
+
+    @Override
+    public Map<String, ComponentTemplate> getComponentTemplateSubstitutions() throws IOException {
+        if (componentTemplateSubstitutions == null) {
+            return Map.of();
+        }
+        Map<String, ComponentTemplate> result = new HashMap<>(componentTemplateSubstitutions.size());
+        for (Map.Entry<String, Map<String, Object>> rawEntry : componentTemplateSubstitutions.entrySet()) {
+            result.put(rawEntry.getKey(), convertRawTemplateToComponentTemplate(rawEntry.getValue()));
+        }
+        return result;
+    }
+
+    @SuppressWarnings("unchecked")
+    private static ComponentTemplate convertRawTemplateToComponentTemplate(Map<String, Object> rawTemplate) throws IOException {
+        Settings settings = null;
+        CompressedXContent mappings = null;
+        if (rawTemplate.containsKey("mappings")) {
+            mappings = new CompressedXContent((Map<String, Object>) rawTemplate.get("mappings"));
+        }
+        if (rawTemplate.containsKey("settings")) {
+            settings = Settings.builder().loadFromMap((Map<String, ?>) rawTemplate.get("settings")).build();
+        }
+        Map<String, AliasMetadata> aliases = null;
+        DataStreamLifecycle lifecycle = null;
```

Comment:
Is this going to cause problems with it having unconsumed xcontent? Should we be throwing an exception for all other keys other than `mappings` and `settings` to show the user that those cannot be simulated? Or should we keep the leniency?

Reply:
For example, if we're ignoring eventual data_stream contents, then eventually from https://github.com/elastic/elasticsearch/issues/98877 which wants to:

> Figure out how to specify tsdb settings in component templates. For example index.routing_path can be specified in a composable index template if data stream template' index_mode is set to time_series. 

Then we may need the whole thing for validation? This may be too premature, I'm just curious right now. Would it make sense to just load the whole thing into a `Template` from the JSON?

Reply:
Yeah I guess there's no harm in just using it all. Plus it simplifies my code. At this point though, I'm working with a Map -- think I ought to just pass it through all the way as xcontent? It's not really performance-sensitive since this is simulate code. And it's convenient for the sake of testing to have these overrides as a Map.

Reply:
The benefit is that if `ComponentTemplate.parse` ever changes to do some kind of validation, this would pick that up at least